Frameworks/CoreGraphics/CGPath.mm

```diff
@@ -655,13 +637,46 @@ void CGPathAddRoundedRect(
     UNIMPLEMENTED();
 }
 
+int _CGPathPointCountForElementType(CGPathElementType type) {
+    int pointCount = 0;
+
+    switch (type) {
+        case kCGPathElementMoveToPoint:
+        case kCGPathElementAddLineToPoint:
+            pointCount = 1;
+            break;
+        case kCGPathElementAddQuadCurveToPoint:
+            pointCount = 2;
+            break;
+        case kCGPathElementAddCurveToPoint:
+            pointCount = 3;
+            break;
+        case kCGPathElementCloseSubpath:
+            pointCount = 0;
+            break;
+    }
+    return pointCount;
+}
+
 /**
  @Status Interoperable
 */
 void CGPathApply(CGPathRef path, void* info, CGPathApplierFunction function) {
+    if (path == NULL) {
+        return;
+    }
+
     for (unsigned i = 0; i < path->_count; i++) {
-        CGPathElement c = path->_components[i];
-        function(info, &c);
+        CGPathElement element = path->_components[i];
+        CGPathElement returnElement;
+        returnElement.type = element.type;
+        int pointCount = _CGPathPointCountForElementType(element.type);
+        returnElement.points = (CGPoint*)IwCalloc(pointCount, sizeof(CGPoint));
+        memcpy(returnElement.points, element.points, pointCount * sizeof(CGPoint));
```

Comment:
We are interpreting "can examine (but not modify) the element" differently.  IMO, it means that the function can examine it, but it should not modify it.  There is really no utility in allowing the function to modify it but guarding against it.  Looking at the applier function: typedef void (*CGPathApplierFunction) ( void *info, const CGPathElement *element );  the const on second parameter implies its contents must not be modified by the function.  We are not really concerned about a rogue user or code here as this runs in the context of the app.  Is there a legitimate case for an app modifying this element? if not, I'd rather remove all these duplications - it will matter a lot for perf.

---
In reply to: [68664667](https://github.com/Microsoft/WinObjC/pull/559#discussion_r68664667) [](ancestors = 68664667)